akka-actor/src/main/java/akka/japi/pf/Reflect.java

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.japi.pf;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Objects;
+
+import sun.reflect.ConstantPool;
+
+/**
+ * INTERNAL API
+ */
+public class Reflect {
+    private static final Method getConstantPoolMethod;
+
+    static {
+        try {
+            getConstantPoolMethod = Class.class.getDeclaredMethod("getConstantPool");
+            getConstantPoolMethod.setAccessible(true);
+        } catch (NoSuchMethodException e) {
+            throw new RuntimeException("Couldn't find the 'getConstantPool' method - are you using the SUN JVM?",
+                                       e);
+        }
+    }
+
+    /**
+     * Resolve the generic argument types of a lamb instance (even though types are not reified in Java)<br/>
+     * Example: If a class/method is handed a Function&lt;Foo, Bar> then resolveGenericLambdaArgumentType for argument 0 (the only argument) will return String.class:
+     * <pre>
+     * Function&lt;String, Boolean> myStrToBoolFunc = strArg -> true;
+     * assertThat(Reflect.resolveGenericLambdaArgumentType(myStrToBoolFunc, 0), equalTo(String.class));
+     * </pre>
+     * Example of a lambda with multiple arguments:
+     * <pre>
+     * BiConsumer&lt;Integer, Long> biConsumer = (i, l) -> {};
+     * assertThat(Reflect.resolveGenericLambdaArgumentType(biConsumer, 0), equalTo(Integer.class));
+     * assertThat(Reflect.resolveGenericLambdaArgumentType(biConsumer, 1), equalTo(Long.class));
+     * </pre>
+     */
+    public static Class<?> resolveLambdaArgumentType(Object object, int argumentIndex) {
+        Objects.requireNonNull(object,
+                               "object");
+        if (argumentIndex < 0) {
+            throw new IllegalArgumentException("ArgumentIndex must be >= 0");
+        }
+
+        Type genericInterface = object.getClass()
+                                      .getGenericInterfaces()[0];
+        if (genericInterface instanceof ParameterizedType) {
+            Type type = ((ParameterizedType) genericInterface).getActualTypeArguments()[argumentIndex];
+            if (type instanceof ParameterizedType) {
+                return (Class<?>) ((ParameterizedType)type).getRawType();
+            } else {
+                return (Class<?>) type;                
+            }
+        }
+        else {
+            String[] methodRef = resolveMethodRef(object, "Failed to resolve generic type arguments for " + object + " of type " + genericInterface.toString());
+            try {
+                if (methodRef[1].startsWith("lambda")) {
+                    // This is an inline lambda expression, e.g. (String s) -> s.isEmpty()
+                    String argumentType = jdk.internal.org.objectweb.asm.Type.getArgumentTypes(methodRef[2])[argumentIndex].getClassName();
```

Comment:
hm, are you sure this isn't reachable using plain reflection?
// please don't use internal JDK api, we won't be able to upgrade to JDK9 with such thingies :)